### PR TITLE
Update vtkMRMLSliceNode to use MRML helper macros

### PR DESCRIPTION
**Pull request by stephencrowell**
`vtkMRMLSliceNode`'s read and write XML did not use macros. As I was updating the XML attributes as part of #7021, we decided to update all XML attributes before adding the new items.

### CODE REVIEW

**Review by pieper — DISMISSED** [review]
Makes sense to me to clean up this code.  Is there test coverage for this?

**Comment by jcfr** [comment]
To provide some context, I suggested @stephencrowell to work on this after realizing the class was still using the "legacy" and more verbose approach.

Once this is done, the plan is to rebase and update the following pull request:
* #7021

File: Libs/MRML/Core/vtkMRMLSliceNode.cxx [hunk 1]
```diff
@@ -858,62 +858,31 @@ void vtkMRMLSliceNode::WriteXML(ostream& of, int nIndent)
...
-  of << " sliceToRAS=\"" << ss.str().c_str() << "\"";
-
+  vtkMRMLWriteXMLBeginMacro(of);
+  vtkMRMLWriteXMLStdFloatVectorMacro(fieldOfView, FieldOfView, double);
+  vtkMRMLWriteXMLStdIntVectorMacro(dimensions, Dimensions, int);
+  vtkMRMLWriteXMLStdFloatVectorMacro(xyzOrigin, XYZOrigin, double);
+  vtkMRMLWriteXMLStringMacro(sliceResolutionMode, SliceResolutionMode);
+  vtkMRMLWriteXMLStdFloatVectorMacro(uvwExtents, UVWExtents, double);
+  vtkMRMLWriteXMLStdIntVectorMacro(uvwDimensions, UVWDimensions, int);
+  vtkMRMLWriteXMLStdFloatVectorMacro(uvwDimensions, UVWOrigin, double);
```
Comment (jcfr) [inline Libs/MRML/Core/vtkMRMLSliceNode.cxx:868]:
```suggestion
  vtkMRMLWriteXMLStdFloatVectorMacro(uvwOrigin, UVWOrigin, double);
```

File: Libs/MRML/Core/vtkMRMLSliceNode.cxx [hunk 2]
```diff
@@ -858,62 +858,31 @@ void vtkMRMLSliceNode::WriteXML(ostream& of, int nIndent)
...
+  vtkMRMLWriteXMLStringMacro(defaultOrientation, DefaultOrientation);
+  vtkMRMLWriteXMLStringMacro(orientationReference, OrientationReference);
+  vtkMRMLWriteXMLIntMacro(jumpMode, JumpMode);
+  vtkMRMLWriteXMLIntMacro(sliceVisibility, SliceVisible);
+  vtkMRMLWriteXMLIntMacro(widgetVisibility, WidgetVisible);
+  vtkMRMLWriteXMLIntMacro(widgetVisibility, WidgetOutlineVisible);
+  vtkMRMLWriteXMLIntMacro(useLabelOutline, UseLabelOutline);
+  vtkMRMLWriteXMLIntMacro(sliceSpacingMode, SliceSpacingMode);
+  vtkMRMLWriteXMLStdFloatVectorMacro(prescribedSliceSpacing, PrescribedSliceSpacing, double);
+  vtkMRMLWriteXMLStdStringVectorMacro(threeDViewNodeRef, ThreeDViewIDs, std::string);
```
Comment (jcfr) [inline Libs/MRML/Core/vtkMRMLSliceNode.cxx:883]: To ensure the MRML scene are identically serialized, this should be moved after `orientationMatrix`

File: Libs/MRML/Core/vtkMRMLSliceNode.cxx [hunk 3]
```diff
@@ -858,62 +858,31 @@ void vtkMRMLSliceNode::WriteXML(ostream& of, int nIndent)
...
+  vtkMRMLWriteXMLStringMacro(layoutGridRows, LayoutGridRows);
+  vtkMRMLWriteXMLStringMacro(layoutGridColumns, LayoutGridColumns);
+  vtkMRMLWriteXMLMatrix4x4Macro(sliceToRAS, SliceToRAS);
+  vtkMRMLWriteXMLStdStringMacro(orientation, Orientation);
+  vtkMRMLWriteXMLStringMacro(defaultOrientation, DefaultOrientation);
+  vtkMRMLWriteXMLStringMacro(orientationReference, OrientationReference);
+  vtkMRMLWriteXMLIntMacro(jumpMode, JumpMode);
+  vtkMRMLWriteXMLIntMacro(sliceVisibility, SliceVisible);
+  vtkMRMLWriteXMLIntMacro(widgetVisibility, WidgetVisible);
+  vtkMRMLWriteXMLIntMacro(widgetVisibility, WidgetOutlineVisible);
```
Comment (jcfr) [inline Libs/MRML/Core/vtkMRMLSliceNode.cxx:879]:
```suggestion
  vtkMRMLWriteXMLIntMacro(widgetOutlineVisibility, WidgetOutlineVisible);
```

**Review by jcfr — CHANGES_REQUESTED** [review]
Thanks for working on this :pray: 

Before we can integrate:
* Consider using `vtkMRML*XMLBooleanMacro` where it applies
* Ensure that order of attribute remains the same
* Locally run the related tests `ctest -R vtkMRMLSlice`

**Comment by stephencrowell** [comment]
I added the macros to `Copy` and `PrintSelf` while we were making changes.

The local tests passed without issue when ran locally.

**Comment by jcfr** [comment]
@stephencrowell To streamline the review process, I ended up squashing the commit together.

**Comment by jcfr** [comment]
Comparing the scene before/after:

image

**Comment by stephencrowell** [comment]
> Comparing the scene before/after:

Looks like the previous version just put the enum value where I used enum macro. I'll change it the int macro to match the previous.

**Comment by jcfr** [comment]
Differences associated with the output of `print(getNode("vtkMRMLSliceNodeRed"))`:

```diff
diff --git a/print.txt b/print.txt
index 711ae5d..23f2d7b 100644
--- a/print.txt
+++ b/print.txt
@@ -27,27 +27,21 @@ vtkMRMLSliceNode (0xFFFFFFF)
   RulerType: none
   RulerColor: white
    AxisLabels: L;R;P;A;I;S
-  FieldOfView:
-      561.404     256     1
-  Dimensions:
-      500     228     1
-  XYZOrigin:
-      0     0     0
-  UVWDimensions:
-      256     256     1
-  UVWExtents:
-      561.404     256     1
-  UVWOrigin:
-      0     0     0
+  FieldOfView: (561.404, 256, 1)
+  Dimensions: (500, 228, 1)
+  XYZOrigin: (0, 0, 0)
+  UVWDimensions: (256, 256, 1)
+  UVWExtents: (561.404, 256, 1)
+  UVWOrigin: (0, 0, 0)
   SliceResolutionMode: 1
   Layout grid: 1x1
-  Active slice: 0
+  ActiveSlice: 0
   SliceVisible: false
   WidgetVisible: false
   WidgetOutlineVisible: true
   UseLabelOutline: false
   Jump mode:   Offset
-  SliceToRAS: 
+  SliceToRAS:
     Debug: Off
     Modified Time: 999999
     Reference Count: 1
@@ -94,7 +88,7 @@ vtkMRMLSliceNode (0xFFFFFFF)
         0 1.12281 0 -121.571 
         0 0 1 -10.2143 
         0 0 0 1 
-  Slice spacing mode: Automatic
-  Prescribed slice spacing: (1, 1, 1)
+  SliceSpacingMode: Automatic
+  PrescribedSliceSpacing: (1, 1, 1)
   Interacting: off
   DefaultOrientation: Axial

```

**Comment by jcfr** [comment]
In term of MRML scene saving, the only difference is one space when serializing `sliceToRAS`:

image

**Comment by jcfr** [comment]
@Sunderlandkyl When you introduced `vtkMRMLWriteXMLMatrix4x4Macro` in e234c65d9a, did you intended separating each "row" with two spaces ? Or was this an oversight ?

**Comment by Sunderlandkyl** [comment]
I don't recall. It was likely unintentional.

**Comment by jcfr** [comment]
Thanks for the follow up, I will update the macro to have only one space then.

Since there are no mrml scenes with the `interactionHandleToWorldMatrix` checked in the source tree (and many with `sliceToRAS` attribute), it will be easy to address.

Also worth noting that the reading of the attribute values should not be impacted.

**Review by jcfr — APPROVED** [review]

**Comment by jcfr** [comment]
After locally rebuilding, I confirm the MRML scene are identical:

image


And different with the output `PrintSelf` is also acceptable:

```diff
diff --git a/print.txt b/print.txt
index 711ae5d..23f2d7b 100644
--- a/print.txt
+++ b/print.txt
@@ -27,27 +27,21 @@ vtkMRMLSliceNode (0xFFFFFFF)
   RulerType: none
   RulerColor: white
    AxisLabels: L;R;P;A;I;S
-  FieldOfView:
-      561.404     256     1
-  Dimensions:
-      500     228     1
-  XYZOrigin:
-      0     0     0
-  UVWDimensions:
-      256     256     1
-  UVWExtents:
-      561.404     256     1
-  UVWOrigin:
-      0     0     0
+  FieldOfView: (561.404, 256, 1)
+  Dimensions: (500, 228, 1)
+  XYZOrigin: (0, 0, 0)
+  UVWDimensions: (256, 256, 1)
+  UVWExtents: (561.404, 256, 1)
+  UVWOrigin: (0, 0, 0)
   SliceResolutionMode: 1
   Layout grid: 1x1
-  Active slice: 0
+  ActiveSlice: 0
   SliceVisible: false
   WidgetVisible: false
   WidgetOutlineVisible: true
   UseLabelOutline: false
   Jump mode:   Offset
-  SliceToRAS: 
+  SliceToRAS:
     Debug: Off
     Modified Time: 999999
     Reference Count: 1
@@ -94,7 +88,7 @@ vtkMRMLSliceNode (0xFFFFFFF)
         0 1.12281 0 -121.571 
         0 0 1 -10.2143 
         0 0 0 1 
-  Slice spacing mode: Automatic
-  Prescribed slice spacing: (1, 1, 1)
+  SliceSpacingMode: Automatic
+  PrescribedSliceSpacing: (1, 1, 1)
   Interacting: off
   DefaultOrientation: Axial
```

**Pull request merged**